lib/CL/devices/cuda/pocl-cuda.c

```diff
@@ -1440,6 +1459,53 @@ pocl_cuda_submit_node (_cl_command_node *node, cl_command_queue cq, int locked)
                                 cmd->memfill.pattern,
                                 cmd->memfill.pattern_size);
       break;
+    case CL_COMMAND_SVM_MAP:
+    case CL_COMMAND_SVM_UNMAP:
+      /* empty */
+      break;
+
+    case CL_COMMAND_SVM_MEMCPY:
+      pocl_cuda_svm_copy_async (stream, cmd->svm_memcpy.dst,
+                                cmd->svm_memcpy.src, cmd->svm_memcpy.size);
+      break;
+    case CL_COMMAND_SVM_MEMFILL:
+      pocl_cuda_submit_memfill (stream, cmd->svm_fill.svm_ptr,
+                                cmd->svm_fill.size, 0, cmd->svm_fill.pattern,
+                                cmd->svm_fill.pattern_size);
+      break;
+    case CL_COMMAND_SVM_FREE:
+      if (cmd->svm_free.pfn_free_func)
+        {
+          cmd->svm_free.pfn_free_func (
+              cmd->svm_free.queue, cmd->svm_free.num_svm_pointers,
+              cmd->svm_free.svm_pointers, cmd->svm_free.data);
+        }
+      else
+        {
+          int i;
+          for (i = 0; i < cmd->svm_free.num_svm_pointers; i++)
+            {
+              void *ptr = cmd->svm_free.svm_pointers[i];
+              POCL_LOCK_OBJ (event->context);
+              pocl_svm_ptr *tmp = NULL, *item = NULL;
+              DL_FOREACH_SAFE (event->context->svm_ptrs, item, tmp)
+              {
+                if (item->svm_ptr == ptr)
+                  {
+                    DL_DELETE (event->context->svm_ptrs, item);
+                    break;
+                  }
+              }
+              POCL_UNLOCK_OBJ (event->context);
+              assert (item);
+              POCL_MEM_FREE (item);
+              // Leads to 'undefined symbol: POclReleaseContext'
+              // POname (clReleaseContext) (event->context);
+
+              dev->ops->svm_free (dev, ptr);
```

Comment:
This free should be stream-synchronized.

Reply:
To add a bit more info, we tried to stream-synchronize the free via `cuMemFreeAsync` (https://github.com/pocl/pocl/pull/1067/commits/bf1642ea34db1236e7e8796d3ba0a09c3a11f94b), but it turns out `cuMemFreeAsync` does not work with managed allocations (it returns `CUDA_ERROR_NOT_SUPPORTED`). 

We could do 2 things:

- Synchronize in a different way (how?)
- Don't use managed memory

Reply:
Since we're only looking to have coarse-grain buffer SVM on Nvidia, I think moving away from managed memory might make sense.